src/jailer/src/env.rs

```diff
@@ -168,20 +171,34 @@ impl Env {
 
         SyscallReturnCode(unsafe { libc::chown(dev_path.as_ptr(), self.uid(), self.gid()) })
             .into_empty_result()
-            .map_err(|e| Error::ChangeFileOwner(dev_path.to_str().unwrap(), e))
+            // Safe to unwrap as we provided valid file names.
+            .map_err(|e| Error::ChangeFileOwner(PathBuf::from(dev_path.to_str().unwrap()), e))
     }
 
-    fn setup_jailed_folders(&mut self) -> Result<()> {
-        for folder in FOLDER_HIERARCHY.iter() {
-            let folder_cstr = CStr::from_bytes_with_nul(folder).map_err(Error::FromBytesWithNul)?;
-
-            // This unwrap is safe as we provided strings that have valid utf8 chars.
+    fn setup_jailed_folders<P>(&self, folders: Vec<P>) -> Result<()>
+    where
+        P: AsRef<Path> + AsRef<OsStr>,
+    {
+        for folder in folders.iter() {
+            let mut folder_bytes = Path::new(folder).as_os_str().as_bytes().to_vec();
+            folder_bytes.push(b'\0');
+            let folder_cstr = CStr::from_bytes_with_nul(folder_bytes.as_slice())
+                .map_err(Error::FromBytesWithNul)?;
+
+            // Safe to unwrap as the byte sequence is UTF-8 validated above.
             let path = folder_cstr.to_str().unwrap();
-            fs::create_dir_all(path).map_err(|e| Error::CreateDir(PathBuf::from(path), e))?;
-
-            SyscallReturnCode(unsafe { libc::chown(folder_cstr.as_ptr(), self.uid(), self.gid()) })
+            let path_buf = PathBuf::from(path);
```

Comment:
Most of the [errors](https://github.com/firecracker-microvm/firecracker/blob/608e135299b5ea46a3c7a3f9a240b229b8b03357/src/jailer/src/main.rs#L27) contain a `PathBuf`, the cloning is for convenience. A `Path` slice would have to be passed on with lifetime and all and it didn't seem quite worth the effort.

Reply:
I was referring to the `path` variable, from the same scope, which is a `str` slice. It can be used instead of the newly created `path_buf` in every place where `path` and `path_buf` are used, without a lifetime modifier. I've built it  without a lifetime modifier.

Reply:
Yup I changed it back to `&str` but the error types still take a `PathBuf` so unless we change *all* of those (which I'd rather not do, at least not in this PR), some transformation is still necessary in thr `map_err` to make a `PathBuf` from a `&str`... unless I'm completely missing something, in which case, please help :grinning: The way I see it now, wither a `PathBuf` is precreated and `clone`d, or recreated at every `map_err` with `path.into()`. I think both options cost pretty much the same amount of computation.

Reply:
Hmm, first time I suggested the change, I've managed to make it compile with `&str` everywhere. It seems it doesn't work at the moment. You can ignore my comment.

Reply:
Generally, we shouldn't jump through hoops to optimize error cases. Those can be slow 😄